### PR TITLE
[C++20] Destroying delete can cause a type to be noexcept when deleting

### LINKED ISSUE #118660: Clang considers destructor in noexcept expression that should call destroying delete
When a destroying delete function is called, the pointed-to object's destructor is not run --- it is the responsibility of the destroying delete function to run any destructors.  However, when querying the noexcept operator for such a delete expression, Clang is considering the exception specification of the destructor, even though at runtime it clearly performs the right code.

```
#include <cstdio>
#include <new>

struct Explicit {
    ~Explicit() noexcept(false) { puts("should not be called");}

    void operator delete(Explicit*, std::destroying_delete_t) noexcept {
        std::puts("destroyed explicit");
    }
};

// #define  SHOW_BUGS
#if defined(SHOW_BUGS)

Explicit * qn = nullptr;
static_assert( noexcept(delete(qn)));

#endif

int main() {
    Explicit *q = new Explicit();
    delete q;
}
```

Here us a Compiler Explorer link to a more complete test, that passed on EDG even with the SHOW_BUGS macro defined:
   https://godbolt.org/z/YaaeT8W17

### PR DESCRIPTION
**Pull request by AaronBallman**
Given a `noexcept` operator with an operand that calls `delete`, Clang was not considering whether the selected `operator delete` function was a destroying delete or not when inspecting whether the deleted object type has a throwing destructor. Thus, the operator would return `false` for a type with a potentially throwing destructor even though that destructor would not be called due to the destroying delete. Clang now takes the kind of delete operator into consideration.

Fixes #118660

### CODE REVIEW

**Review by cor3ntin — APPROVED** [review]

**Review by erichkeane — APPROVED** [review]

File: clang/test/SemaCXX/noexcept-destroying-delete.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -fcxx-exceptions -Wno-unevaluated-expression -std=c++20 %s
```
Comment (shafik) [inline clang/test/SemaCXX/noexcept-destroying-delete.cpp:1]: We already have 

https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/expr/expr.unary/expr.unary.noexcept/sema.cpp

maybe the test is better there and wrapped in a `GH118660` namespace?

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -654,6 +654,9 @@ Bug Fixes in This Version
 - Fixed a crash when GNU statement expression contains invalid statement (#GH113468).
 - Fixed a failed assertion when using ``__attribute__((noderef))`` on an
   ``_Atomic``-qualified type (#GH116124).
+- No longer return ``false`` for ``noexcept`` expressions involving a
+  ``delete`` which resolves to a destroying delete but the type of the object
+  being deleted as a potentially throwing destructor (#GH118660).
```
Comment (zygoloid) [inline clang/docs/ReleaseNotes.rst:659]:
```suggestion
  being deleted has a potentially throwing destructor (#GH118660).
```
Typo?

Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst:659]: Good catch, fixed.

File: clang/lib/Sema/SemaExceptionSpec.cpp [hunk 3]
```diff
@@ -1206,15 +1206,16 @@ CanThrowResult Sema::canThrow(const Stmt *S) {
       CT = CT_Dependent;
     } else {
       const FunctionDecl *OperatorDelete = DE->getOperatorDelete();
-      CT = canCalleeThrow(*this, DE, OperatorDelete);
-      if (const RecordType *RT = DTy->getAs<RecordType>()) {
-        const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
-        const CXXDestructorDecl *DD = RD->getDestructor();
-        if (DD && !OperatorDelete->isDestroyingOperatorDelete())
+      if (!OperatorDelete->isDestroyingOperatorDelete()) {
+        CT = canCalleeThrow(*this, DE, OperatorDelete);
```
Comment (zygoloid) [inline clang/lib/Sema/SemaExceptionSpec.cpp:1210]:
```suggestion
      CT = canCalleeThrow(*this, DE, OperatorDelete);
      if (!OperatorDelete->isDestroyingOperatorDelete()) {
```
A destroying operator delete itself can still throw, I think.

Comment (AaronBallman) [reply clang/lib/Sema/SemaExceptionSpec.cpp:1210]: Ah, good point! I'll add test coverage for that too.

File: clang/test/SemaCXX/noexcept-destroying-delete.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/test/SemaCXX/noexcept-destroying-delete.cpp:1]: That test file has a lot of dynamic exception specification code in it, so it's leaning heavily on `-std=c++11`.

**Review by erichkeane — APPROVED** [review]

File: clang/lib/Sema/SemaExceptionSpec.cpp [hunk 4]
```diff
@@ -1205,15 +1205,18 @@ CanThrowResult Sema::canThrow(const Stmt *S) {
...
-      CT = canCalleeThrow(*this, DE, DE->getOperatorDelete());
-      if (const RecordType *RT = DTy->getAs<RecordType>()) {
-        const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
-        const CXXDestructorDecl *DD = RD->getDestructor();
-        if (DD)
-          CT = mergeCanThrow(CT, canCalleeThrow(*this, DE, DD));
+      const FunctionDecl *OperatorDelete = DE->getOperatorDelete();
+      CT = canCalleeThrow(*this, DE, OperatorDelete);
+      if (!OperatorDelete->isDestroyingOperatorDelete()) {
+        if (const RecordType *RT = DTy->getAs<RecordType>()) {
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Sema/SemaExceptionSpec.cpp]: Good call; I simplified along these lines

**Review by erichkeane — APPROVED** [review]

**Comment by zygoloid** [comment]
Hm, the issue raised in #118800 is relevant here too, isn't it? If the destructor is virtual, we want to look at its exception specification, not the one on the destroying operator delete, because `delete p` actually invokes the virtual deleting destructor, not the statically selected deallocation function.

Maybe it's best to handle that in #118800, but I think it basically means the changes we made here aren't right :(

**Pull request merged**